### ISSUE TITLE
Update hangs on zsync.

### ISSUE THREAD

**Issue by Biep** [koreader/koreader#5367]
* KOReader version: v2019.08
* Device: Kobo Aura One

#### Issue
When trying to update, the procedure stalls while computing zsync.  The activity indicator keeps moving, but waiting several times for 3-6 hours yields nothing.

#### Steps to reproduce
Click "check for update", and accept the update.

##### `crash.log` (if applicable)
I hope this is the relevant part.  The complete file has been attached too.

09/12/19-13:58:15 INFO  Kobo WiFi: enabling WiFi
udhcpc (v1.22.0.kobo) started
Sending discover...
Sending select for 192.168.2.77...
Lease of 192.168.2.77 obtained, lease time 86400
adding dns 192.168.2.254
PING 192.168.2.254 (192.168.2.254): 56 data bytes
64 bytes from 192.168.2.254: seq=0 ttl=64 time=13.809 ms

--- 192.168.2.254 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 13.809/13.809/13.809 ms
Original fb rotation is set @ 0
Original fb bitdepth is set @ 16bpp
Switching fb bitdepth to 8bpp & rotation to Portrait
[FBInk] Detected a Kobo Aura One (373 => Daylight @ Mark 6)
Variable fb info: 1872x1404 (1888x2816), 16bpp @ rotation: 0 (Upright, 0°)
Fixed fb info: ID is "mxc_epdc_fb", length of fb mem: 10813440 bytes & line length: 3776 bytes
Device's expected Portrait orientation should be: 3 (Counter Clockwise, 270°)!

Switching fb to 8bpp @ rotation 3 . . .
Setting bitdepth to 8bpp
Setting grayscale to 1
Setting rotate to 3 (Counter Clockwise, 270°)
Bitdepth is now 8bpp (grayscale: 1) @ rotate: 3 (Counter Clockwise, 270°)

Variable fb info: 1404x1872 (1408x3840), 8bpp @ rotation: 3 (Counter Clockwise, 270°)
Fixed fb info: ID is "mxc_epdc_fb", length of fb mem: 10813440 bytes & line length: 1408 bytes
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 09/12/19-16:14:59
 [*] Version: v2019.08

[crash.log](https://github.com/koreader/koreader/files/3605941/crash.log)

**Comment by Biep** [comment]
> If you have a *nix box connected to the same network, running zsync from there as an experiment could be interesting ;).

I have an Ubuntu machine, and am quite willing to help test things - but you'll have to guide me step by step, because I hardly know what I am doing.

**Comment by Frenzie** [comment]
I don't know what NiLuJe has in mind, but you could see if the four commands I posted above show a difference. Something like this.

Run in shell:
```shell
# Install zsync.
sudo apt install zsync
```

Save as script and run the script:
```shell
#!/bin/sh

# Test OTA server one.
echo "Stable download from scratch:"
time zsync -o bla http://koreader.ak-team.com/koreader-kobo-latest-stable.zsync
echo "Update to nightly from stable:"
time zsync -o bla http://koreader.ak-team.com/koreader-kobo-latest-nightly.zsync

# Test OTA server two.
echo "Stable download from scratch:"
time zsync -o bla2 http://ota.koreader.rocks/koreader-kobo-latest-stable.zsync
echo "Update to nightly from stable:"
time zsync -o bla2 http://ota.koreader.rocks/koreader-kobo-latest-nightly.zsync

# Remove test files.
rm bla
rm bla2
```

The delta has now increased quite a bit compared to the stable, unlike yesterday. `v2019.09.3-7-g840d05b` has several new libraries, which should carry a lot more weight than a few changed text files.

There was an extreme drop in performance on the CDN. Or presumably it would just be more accurate to say that yesterday the amount of data to be downloaded was so small that the speed hardly mattered at all.

```shell
# ak-team
real	0m3.743s
user	0m1.641s
sys	0m0.178s

real	0m1.128s
user	0m0.434s
sys	0m0.057s


# ota.koreader.rocks
# It still showed speeds going up to 10 MB/s
# but I guess it took a little longer to get going.
real	0m6.088s
user	0m1.558s
sys	0m0.186s

# Wow… it's supposed to save time by only downloading a few hundred kB to a few MB
# but instead it would've been faster to just download the whole thing.
# Of course on H2O the potential top speed is about 1.5 MB/s, not 10 MB/s.
real	0m15.847s
user	0m0.436s
sys	0m0.106s
```

It looks like I was effectively largely wrong in https://github.com/koreader/koreader/issues/2040#issuecomment-480828442 but didn't notice because the H2O is pretty slow at this to begin with. Basically its I/O can drop to 100-150 kB/s all by itself when dealing with less linear stuff, without any help from zsync/Cloudflare.

**Comment by Biep** [comment]
```
Stable download from scratch:
#################### 100.0% 10.9 kBps DONE    

No relevent local data found - I will be downloading the whole file. If that's not what you want, CTRL-C out. You should specify the local file is the old version of the file to download with -i (you might have to decompress it with gzip -d first). Or perhaps you just have no data that helps download the file
downloading from http://koreader.ak-team.com/koreader-kobo-v2019.09.3.targz:
#################### 100.0% 500.1 kBps DONE     

verifying download...checksum matches OK
used 0 local, fetched 35265515
4.34user 1.64system 1:10.09elapsed 8%CPU (0avgtext+0avgdata 4888maxresident)k
0inputs+139896outputs (0major+968minor)pagefaults 0swaps
Update to nightly from stable:
#################### 100.0% 9.6 kBps DONE     

Read bla. Target 83.8% complete.      ********************************************************
downloading from http://koreader.ak-team.com/koreader-kobo-v2019.09.3-13-g2a93e42_2019-09-15.targz:
#################### 100.0% 284.0 kBps DONE     

verifying download...checksum matches OK
used 59764736 local, fetched 3573958
1.06user 0.34system 0:13.44elapsed 10%CPU (0avgtext+0avgdata 4796maxresident)k
0inputs+140056outputs (0major+977minor)pagefaults 0swaps
Stable download from scratch:
#################### 100.0% 162.5 kBps DONE     

No relevent local data found - I will be downloading the whole file. If that's not what you want, CTRL-C out. You should specify the local file is the old version of the file to download with -i (you might have to decompress it with gzip -d first). Or perhaps you just have no data that helps download the file
downloading from http://ota.koreader.rocks/koreader-kobo-v2019.09.3.targz:
#################### 100.0% 544.4 kBps DONE     

verifying download...checksum matches OK
used 0 local, fetched 35265651
4.27user 1.42system 1:06.09elapsed 8%CPU (0avgtext+0avgdata 4832maxresident)k
0inputs+139904outputs (0major+966minor)pagefaults 0swaps
Update to nightly from stable:
#################### 100.0% 163.0 kBps DONE     

Read bla2. Target 83.8% complete.      ********************************************************
downloading from http://ota.koreader.rocks/koreader-kobo-v2019.09.3-13-g2a93e42_2019-09-15.targz:
#################### 100.0% 150.0 kBps DONE     

verifying download...checksum matches OK
used 59764736 local, fetched 3575874
1.23user 0.32system 0:25.77elapsed 6%CPU (0avgtext+0avgdata 4812maxresident)k
40inputs+140056outputs (0major+981minor)pagefaults 0swaps
```

**Comment by protist** [comment]
I just tried to update to the latest release OTA on my Kobo H2O. All previous updates have worked perfectly fine, but v2020.01 is stalling for me at the `Computing zsync delta` stage, with the spinner still rotating. I've tried and failed three times now. I don't have any logs yet, but the worst part is that there doesn't seem to be an obvious way to cancel it.

The first time, I *think* that holding/pressing the hard button slept the device. Upon waking I had the "failed" screen. The second time and third time, holding the button did nothing. The second time, I had to leave it on all night until I got the "battery low" dialogue box, which I could not dismiss, but at least it freed the hard button for a hard restart. I'm now stuck again, but guess I'll have to wait a day or two for the battery to die again! Would it be possible to introduce a way to cancel an in-progress OTA update?

**Comment by Frenzie** [comment]
If you press the power button for 5-10 seconds the device will always hard turn off. That can't be prevented by software.

That's really weird about zsync not timing out by itself though… unless it's just the spinner that doesn't know when to give up for some reason?

https://github.com/koreader/koreader/blob/343eff312353b9501ad69f14f366313580f3c6fd/platform/common/spinning_zsync#L15-L43

Probably best to hard-code a timeout of a few mins in any case.

**Comment by protist** [comment]
> If you press the power button for 5-10 seconds the device will always hard turn off. That can't be prevented by software.

That's the weird thing though. It absolutely does not work. I tried holding it down for 30 seconds multiple times! Unless... my power button is starting to fail, but I think that's unlikely because it's worked every time prior to these issues, and then twice more as detailed above.

> Probably best to hard-code a timeout of a few mins in any case.

That sounds like a great solution.

**Comment by Frenzie** [comment]
There's also the reset pinhole on the right of the USB connector. But don't overdo it there; I forget the specifics but you can trigger a full factory reset by spamming it or holding it or something.

**Comment by NiLuJe** [comment]
@Frenzie: The very next line is the subshell being murdered with prejudice as soon as zsync exits ;).

So, AFAICT, either zsync was still busy, or it hung without exiting.

Hopefully this'll be a thing of the past if the zsync2 experiment pans out ;).

----

@protist: Same question as usual, on which OTA server is that?

----

FWIW, update to 2020.01 went smooth as butter on my Forma yesterday, so, not a release-specific weirdness like we've seen before.

**Comment by Frenzie** [comment]
@NiLuJe I noticed, but I don't really know zsync's behavior in case of weirdness.

**Comment by NiLuJe** [comment]
I wouldn't put "hanging for no sensible reason" past it ^^.

**Comment by protist** [comment]
> There's also the reset pinhole on the right of the USB connector. But don't overdo it there; I forget the specifics but you can trigger a full factory reset by spamming it or holding it or something.

@Frenzie I think I read somewhere that even a short press would trigger a full factory reset, so I avoided it just in case. I can also confirm that my power button isn't broken!

> Same question as usual, on which OTA server is that?

@NiLuJe I was also trying http://ota.koreader.rocks . I just tried switching to http://koreader.ak-team.com, but that also failed. I'm not 100% sure of the OTA update process, but this attempt seemed to miss the first step (calculating or downloading or…?), and jumped straight to `Computing zsync delta`. I'm not sure if that is important. Also, I started around 6:39 AM UTC if that helps for the server logs.

**Comment by Frenzie** [comment]
> I think I read somewhere that even a short press would trigger a full factory reset
 
I don't think so. :-)

https://help.kobo.com/hc/en-us/articles/360017605234
> Insert a paper clip into the small hole beside the Micro USB slot and push gently. 
> …
> Repeat step 5 three more times. Your eReader's screen will turn black, and show a Critical error message. Don't worry, this is normal.

Btw,
> I can also confirm that my power button isn't broken!

What exactly do you mean by that? That you were able to interrupt the OTA update process?

**Comment by protist** [comment]
> I don't think so. :-)

Ah right. Thanks! I actually skimmed that page and totally misunderstood it.

> What exactly do you mean by that?

Oh, sorry. I could *not* interrupt the OTA update process, at least not consistently. When I tried to press the power button during the update, it did nothing (except perhaps once). I was previously suggesting that perhaps my power button had coincidentally broken, but was confirming here that this was not the case.

**Comment by Frenzie** [comment]
(Un?)fortunately I can confirm that it takes about 8 inaccurately counted seconds of pressing the power button to kill the device. As evidenced by the spinner no longer spinning in this specific scenario.

**Comment by NiLuJe** [comment]
@protist: I think I'm seeing you, (I certainly see *someone* with a zsync that painfully chugged along from 4:06 AM to 7:57 AM (GMT +1) ^^).

That seems *a bit* overkill, but it did not actually hang :D.

(And the process does jump straight to "Computing zsync delta", that's normal. The only possible preceding step is generating the local tarball, which gets skipped when it already exists, which is always true except on fresh installs).

**Comment by protist** [comment]
> (Un?)fortunately I can confirm

Thanks @Frenzie! That makes it much easier to troubleshoot. I tested it and it worked perfectly.

> I think I'm seeing you

@NiLuJe Those times don't quite seem to match though? Just in case, I tried it again, this time from 23:59 UTC (31 Jan) to 0:34 UTC (1 Feb).

I have no idea why it would be chugging along slowly. I've actually been having intermittent dropouts in my internet connection, but when it's up, the speed is solid. AFAICT there were no dropouts during this period.

I also had a rummage through the Kobo's file system. `crash.log` was pretty uninformative, as expected, but these sizes might be important?
```
$ ls -Alt KOBOeReader/.adds/koreader/ota 
total 143504
-rw-r--r-- 1 protist protist 72529920 Feb  1 11:01 koreader.updated.tar.part
-rw-r--r-- 1 protist protist 73943040 Feb  1 11:00 koreader.installed.tar
-rw-r--r-- 1 protist protist   425166 Feb  1 10:59 koreader-kobo-latest-stable.zsync
-rw-r--r-- 1 protist protist    43316 Dec 31 03:42 package.index
```

**Comment by NiLuJe** [comment]
Yeah, I think I see you (server clock happens to be fast by a few minutes, my bad ^^).

You're apparently not the only one going through a long slog of zsync requests, but I can't reeally tell much other than that ;).

What I can try is replicating what you were doing (plus, that'll be a good benchmark for my zsync2 tests), so: from which exact version were you updating?

**Comment by protist** [comment]
Thanks @NiLuJe. I'm updating from the previous v2019.12.

**Comment by Frenzie** [comment]
@protist 
> Thanks @Frenzie! That makes it much easier to troubleshoot. I tested it and it worked perfectly.

I'm afraid I don't quite follow? I merely verified that what I said should work does indeed work on my device, which you previously said doesn't work on yours?

**Comment by protist** [comment]
@Frenzie Sorry… I totally misread your comment. I thought you were talking about using a paperclip on the hidden button, which I can confirm works for me. Yes, the power button still doesn't work for me in this particular scenario.

**Comment by NiLuJe** [comment]
Huh.

```
time ../zsync -i installed.tar -o updated.tar -u http://koreader.ak-team.com/ koreader-kobo-latest-stable.zsync                                                                                                                                                                                                              ──(Sat, Feb 01)─┘
#################### 100.0% 3027.1 kBps DONE     

Done reading installed.tar. 64.6% of target obtained.      
#################### 100.0% 5601.7 kBps DONE     

Done reading updated.tar.part. 64.6% of target obtained.      
downloading new blocks from http://koreader.ak-team.com/koreader-kobo-v2020.01.targz:
############-------- 64.6%koreader.ak-team.com: Success
############-------- 64.6% 0.0 kBps aborted    

failed to retrieve from http://koreader.ak-team.com/koreader-kobo-v2020.01.targz
Not all of the required data could be downloaded, and the remaining data could not be retrieved from any of the download URLs. Incomplete transfer left in updated.tar.part.
(If this is the download filename with .part appended, zsync will automatically pick this up and reuse the data it has already done if you retry in this dir.)
- Real: 31516ms User: 14710ms System: 1660ms Percent: 51% Cmd: ../zsync -i installed.tar -o updated.tar -u http://koreader.ak-team.com/  MajFault: 3 VolContSw: 137 InvContSw: 2248 -
```

```
┌─(ROOT@(none):pts/0)─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/mnt/onboard/.adds/koreader/foo)─┐
└─(1.43:17%:19:54:29%:#)── time ../zsync2 -i installed.tar -o updated.tar -u http://koreader.ak-team.com/ koreader-kobo-latest-stable.zsync                                                                                                                                                                                                             ──(Sat, Feb 01)─┘
zsync2 version 2.0.0-alpha-1 (commit f41aed2), build <local dev build> built on 2020-02-01 18:51:35 UTC
Checking for changes...
Cannot find file updated.tar, triggering delta download
Target file: /mnt/onboard/.adds/koreader/foo/updated.tar
Reading seed file: installed.tar
Usable data from seed files: 64.613864%
Renaming temp file
Fetching remaining blocks
Downloading from http://koreader.ak-team.com/koreader-kobo-v2020.01.targz
optimized ranges, old requests count 894, new requests count 7


#############------- 65.9% 334.2 kBps           


###################- 95.5% 887.9 kBps            


###################- 97.2% 5473.8 kBps          


###################- 98.1% 28057.6 kBps         


###################- 98.1% 9580.3 kBps          


###################- 98.1% 10128.0 kBps         


#################### 100.0% 6754.3 kBps DONE    

Verifying downloaded file
checksum matches OK
used 46864384 local, fetched 17430669
- Real: 32647ms User: 12860ms System: 2060ms Percent: 45% Cmd: ../zsync2 -i installed.tar -o updated.tar -u http://koreader.ak-team.com/  MajFault: 0 VolContSw: 3805 InvContSw: 19917 -
```

TL;DR: zsync actually fails on my end, but it does so fairl quickly.

zsync2 works, but I'm tracking down useless temp files writing right now...

**Comment by NiLuJe** [comment]
Oh, zsync also writes a stupid temp file to checksum the seed file -_-".

Which explains why this is always slower than expected on devices, writing 70MB just for fun isn't great...

**Comment by NiLuJe** [comment]
Well, off to deal with building libcurl and messing with CMake for zsync2, then :D.

**Comment by NiLuJe** [comment]
@pazos: Can you confirm that we're never ever using zsync on Android, so I can blissfully ignore that?

**Comment by Frenzie** [comment]
@NiLuJe Either I or @pazos stripped that out of Android a while back because it didn't make much sense. Because while sometimes you can update the Lua files without updating the runtime, that's just asking for trouble every few weeks really (maybe every few months if you're lucky).

**Comment by Frenzie** [comment]
Instead the update check just starts a download of the latest APK build when prompted.

**Comment by NiLuJe** [comment]
Oh, wait. I think I just found out why we're adding `:80` to the server URLs xD.

**Comment by NiLuJe** [comment]
So, not always a win (24 december nightly -> 2020.01):

```
┌─(ROOT@(none):pts/0)─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/mnt/onboard/.adds/koreader/foo)─┐
└─(1.75:17%:20:20:24%:#)── time ../zsync -i installed.tar -o updated.tar -u http://koreader.ak-team.com:80/ koreader-kobo-latest-nightly.zsync                                                                                                                                                                                                          ──(Sat, Feb 01)─┘
#################### 100.0% 18007.2 kBps DONE     

Done reading installed.tar. 96.5% of target obtained.      
downloading new blocks from http://koreader.ak-team.com:80/koreader-kobo-v2020.01.targz:
#################### 100.0% 323.0 kBps DONE     

verifying download...checksum matches OK
used 69961728 local, fetched 983209
- Real: 7830ms User: 4230ms System: 1460ms Percent: 72% Cmd: ../zsync -i installed.tar -o updated.tar -u http://koreader.ak-team.com:80/  MajFault: 0 VolContSw: 220 InvContSw: 2116 -
```

```
┌─(ROOT@(none):pts/0)─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/mnt/onboard/.adds/koreader/foo)─┐
└─(1.39:17%:20:18:25%:#)── time ../zsync2 -i installed.tar -o updated.tar -u http://koreader.ak-team.com/ koreader-kobo-latest-nightly.zsync                                                                                                                                                                                                            ──(Sat, Feb 01)─┘
zsync2 version 2.0.0-alpha-1 (commit f41aed2), build <local dev build> built on 2020-02-01 18:51:35 UTC
Checking for changes...
Cannot find file updated.tar, triggering delta download
Target file: /mnt/onboard/.adds/koreader/foo/updated.tar
Reading seed file: installed.tar
Usable data from seed files: 96.459127%
Renaming temp file
Fetching remaining blocks
Downloading from http://koreader.ak-team.com/koreader-kobo-v2020.01.targz
optimized ranges, old requests count 952, new requests count 17


###################- 96.5% 5.5 kBps         A  


###################- 96.5% 565.2 kBps         


###################- 96.5% 775.2 kBps         


###################- 96.5% 841.0 kBps         


###################- 96.5% 500.0 kBps         A  


###################- 96.5% 1551.7 kBps         


###################- 96.5% 1558.1 kBps         


###################- 96.5% 2206.7 kBps         


###################- 96.5% 2208.4 kBps         


###################- 96.5% 2518.9 kBps         


###################- 98.0% 944.8 kBps           


###################- 99.1% 1683.2 kBps          


###################- 99.1% 11936.4 kBps         


###################- 99.2% 12014.0 kBps         


###################- 99.2% 4112.3 kBps          


###################- 99.2% 4654.5 kBps          


#################### 100.0% 3525.9 kBps DONE    

Verifying downloaded file
checksum matches OK
used 69961728 local, fetched 9408877
- Real: 18820ms User: 6540ms System: 1760ms Percent: 44% Cmd: ../zsync2 -i installed.tar -o updated.tar -u http://koreader.ak-team.com/  MajFault: 4 VolContSw: 1767 InvContSw: 10579 -
```

**Comment by NiLuJe** [comment]
Okay, once the `:80` mystery is taken care of, the 2019.12 -> 2020.01 update indeed stalls at 100% w/ zsync, but it ultimately times out

```
┌─(ROOT@(none):pts/0)─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/mnt/onboard/.adds/koreader/foo)─┐
└─(1.06:17%:20:27:24%:#)── time ../zsync -i installed.tar -o updated.tar -u http://koreader.ak-team.com:80/ koreader-kobo-latest-stable.zsync                                                                                                                                                                                                           ──(Sat, Feb 01)─┘
#################### 100.0% 10309.3 kBps DONE     

Done reading installed.tar. 64.6% of target obtained.      
downloading new blocks from http://koreader.ak-team.com:80/koreader-kobo-v2020.01.targz:
###################- 99.7% 699.3 kBps           

downloading new blocks from http://koreader.ak-team.com:80/koreader-kobo-v2020.01.targz:
###################- 100.0% 0.0 kBps         

downloading new blocks from http://koreader.ak-team.com:80/koreader-kobo-v2020.01.targz:
###################- 100.0% 0.0 kBps         

downloading new blocks from http://koreader.ak-team.com:80/koreader-kobo-v2020.01.targz:
###################- 100.0% 0.0 kBps         

downloading new blocks from http://koreader.ak-team.com:80/koreader-kobo-v2020.01.targz:
###################- 100.0% 0.0 kBps         

downloading new blocks from http://koreader.ak-team.com:80/koreader-kobo-v2020.01.targz:
###################- 100.0%koreader.ak-team.com: Connection timed out
###################- 100.0% 0.0 kBps aborted    

failed to retrieve from http://koreader.ak-team.com:80/koreader-kobo-v2020.01.targz
Not all of the required data could be downloaded, and the remaining data could not be retrieved from any of the download URLs. Incomplete transfer left in updated.tar.part.
(If this is the download filename with .part appended, zsync will automatically pick this up and reuse the data it has already done if you retry in this dir.)
- Real: 43510ms User: 10130ms System: 1570ms Percent: 26% Cmd: ../zsync -i installed.tar -o updated.tar -u http://koreader.ak-team.com:80/  MajFault: 0 VolContSw: 1982 InvContSw: 15482 -
```

I'm willing to take the perf hit if it means not having to deal with zsync ever again ^^.

**Comment by NiLuJe** [comment]
Random reminder that trusting the battery gauge on a Kobo will only lead to pain & confusion:

```
┌─(ROOT@(none):pts/0)─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/mnt/onboard/.adds/koreader/foo)─┐
└─(0.06:19%:20:32:24%:#)──                                                                                                                                                                                                                                                                                                                              ──(Sat, Feb 01)─┘
┌─(ROOT@(none):pts/0)─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/mnt/onboard/.adds/koreader/foo)─┐
└─(0.01:19%:20:33:40%:#)──    
```

(The final % on the left prompt is battery).

TL;DR: It was critical, I plugged it in, now it's magically much higher.

**Comment by NiLuJe** [comment]
TL;DR @Biep: You'll have to update manually. Preferrably after I'm done throwing zsync into a fire ;).

**Comment by Frenzie** [comment]
> I'm willing to take the perf hit if it means not having to deal with zsync ever again ^^.

On anything over let's say ~8 Mbps I doubt zsync saves much time compared to a full download tbh, especially taking into account that whole giant temp file thing.

**Comment by NiLuJe** [comment]
Yep, that is also worth considering...

That said, we'd still have to download/write ~30MB somewhere, but we could do away with the 70MB uncompressed tarball (and the its sibling copy for the installed seed).

**Comment by NiLuJe** [comment]
FWIW:

```
┌─(ROOT@(none):pts/0)─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/mnt/onboard/.adds/koreader/foo)─┐
└─(23:0.59:19%:21:38:85%:#)── time curl http://koreader.ak-team.com/koreader-kobo-v2020.01.targz -o koreader-kobo-v2020.01.tar.gz                                                                                                                                                                                                                       ──(Sat, Feb 01)─┘
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 34.0M  100 34.0M    0     0  1022k      0  0:00:34  0:00:34 --:--:-- 1082k
- Real: 34133ms User: 1040ms System: 3250ms Percent: 12% Cmd: curl http://koreader.ak-team.com/koreader-kobo-v2020.01.targz -o  MajFault: 0 VolContSw: 10183 InvContSw: 14236 -
```

:D

**Comment by Biep** [comment]
> TL;DR @Biep: You'll have to update manually. Preferably after I'm done throwing zsync into a fire ;).

Please let me know when you enjoy the bonfire!

**Comment by Frenzie** [comment]
That's a long-term significant time investment "implement some kind of threading" type of feature.

**Comment by NiLuJe** [comment]
Yeah, zsync2 is a actually implemented mainly as a library, so, someones with a severe enough masochistic tendencies could write the whole thing via FFI instead, but, errm. :D.

**Comment by NiLuJe** [comment]
Currently blindly writing CMake & M4 for the build-system, vainly hoping everything will Just Work(TM) on the first attempt.

Cue rainbows & unicorns :D. :unicorn:

**Comment by protist** [comment]
Thanks for checking all of this. Just so I understand correctly, are you saying that this OTA will likely never work for me, and I shouldn't wait for unicorns? I'm happy to do a manual update, but I was holding out (a) just in case OTA might work at come point and (b) so I could potentially help troubleshoot further.

**Comment by NiLuJe** [comment]
Yup, no chance of unicorns ;).

**Comment by NiLuJe** [comment]
"Fixed" via https://github.com/koreader/koreader/pull/5810
